libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/EnhancedLocationEngineDecorator.java

```diff
@@ -0,0 +1,191 @@
+                                  @NonNull ScheduledExecutorService executorService) {
+    this.locationEngine = locationEngine;
+    this.mapboxNavigator = mapboxNavigator;
+    this.executorService = executorService;
+  }
+
+  @Override
+  public void getLastLocation(@NonNull LocationEngineCallback<LocationEngineResult> callback) throws SecurityException {
+    locationEngine.getLastLocation(callback);
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     @NonNull final LocationEngineCallback<LocationEngineResult> callback,
+                                     @Nullable Looper looper) throws SecurityException {
+    locationEngine.requestLocationUpdates(request, callback, looper);
+  }
+
+  @Override
+  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
+                                     PendingIntent pendingIntent) throws SecurityException {
+    locationEngine.requestLocationUpdates(request, pendingIntent);
+  }
+
+  @Override
+  public void removeLocationUpdates(@NonNull LocationEngineCallback<LocationEngineResult> callback) {
+    locationEngine.removeLocationUpdates(callback);
+  }
+
+  @Override
+  public void removeLocationUpdates(PendingIntent pendingIntent) {
+    locationEngine.removeLocationUpdates(pendingIntent);
+  }
+
+  void onActiveGuidance() {
+    removeLocationUpdates(callback);
+    if (future != null) {
+      future.cancel(false);
+      future = null;
+    }
+  }
+
+  void onFreeDrive() {
+    requestLocationUpdates(obtainLocationEngineRequest(), callback, null);
+    final Handler handler = new Handler(Looper.getMainLooper());
+    future = executorService.scheduleAtFixedRate(new Runnable() {
+      @Override
+      public void run() {
+        if (rawLocation != null) {
+          enhancedLocation = getLocation(new Date(), 0, rawLocation);
+          handler.post(new Runnable() {
+            @Override
+            public void run() {
+              callback.onSuccess(LocationEngineResult.create(enhancedLocation));
+            }
+          });
+        }
+      }
+    }, 1500, 1000, TimeUnit.MILLISECONDS);
+  }
+
+  void onNavigationKilled() {
+    if (future != null) {
+      removeLocationUpdates(callback);
+      future.cancel(false);
+      future = null;
+      executorService.shutdown();
+    }
+  }
+
+  void configure(@NonNull File path, @NonNull String version, @NonNull String host,
+                 @NonNull String accessToken) {
+    OfflineNavigator offlineNavigator =
+      new OfflineNavigator(mapboxNavigator.getNavigator(), version, host, accessToken);
+    String tilePath = new File(path, version).getAbsolutePath();
+    offlineNavigator.configure(tilePath, new OnOfflineTilesConfiguredCallback() {
+      @Override
+      public void onConfigured(int numberOfTiles) {
+        Timber.d("DEBUG: onConfigured %d", numberOfTiles);
+        onFreeDrive();
```

Comment:
Is this callback synchronous?

If it isn't, can we run into a scenario where:
1. the `MapboxNavigation` is initialized and `configure` request is sent
2. in the meantime, active guidance is started
3. `OnOfflineTilesConfiguredCallback` callback returns and starts the free drive

If that's the case, we can introduce a state boolean that's going to be checked in the executor's task before asking for nav-native location.